File: Rnwood.Smtp4dev/ClientApp/HubConnectionManager.ts
import { HubConnection, HubConnectionBuilder, LogLevel } from '@aspnet/signalr';

export default class HubConnectionManager {

    private _connection: HubConnection;
    connected: boolean = false;
    connectedCallback: () => void;
    started: boolean = false;
    error: Error|null = null;

    constructor(url: string, connectedCallback: () => void) {
        this._connection = new HubConnectionBuilder().withUrl(url).configureLogging(LogLevel.Trace).build();
        this._connection.onclose(this.onConnectionClosed.bind(this));
        this.connectedCallback = connectedCallback;
        //this._connection.serverTimeoutInMilliseconds = 5000;
    }

    async start() {

        if (this.started) {
            return;
        }

        this.error = null;
        this.started = true;
        try {
            await this._connection.start();
            this.connected = true;
            this.connectedCallback();
        } catch (e) {
            this.error = e;
        }
    }

    async stop() {
        this.started = false;
        await this._connection.stop();
    }

    onConnectionClosed(e: Error|undefined) {
        this.connected = false;
        this.started = false;
        this.error = e || null;
    }


    on(eventName: string, handler: (...args: any[]) => void) {
        this._connection.on(eventName, handler);
    }



}

File: Rnwood.Smtp4dev.Client/HubConnectionManager.cs
using Microsoft.AspNetCore.SignalR.Client;

namespace Rnwood.Smtp4dev.Client
{
    public class HubConnectionManager
    {
        private HubConnection _connection;
        public bool Connected { get; private set; } = false;

        public bool Started { get; private set; } = false;

        public Exception Error { get; private set; }

        private List<Func<Task>> connectedCallbacks = new List<Func<Task>>();


        public HubConnectionManager(string url)
        {
            this._connection = new HubConnectionBuilder().WithUrl(url).Build();

            this._connection.Closed += ConnectionClosed;
            this._connection.Reconnected += Connection_Reconnected;
        }

        private async Task ConnectionClosed(Exception e)
        {
            this.Connected = false;
            this.Started = false;
            this.Error = e;       
            await OnStatusChanged();
        }

        private async Task Connection_Reconnected(string arg)
        {
            foreach (var callback in connectedCallbacks)
            {
                await callback();
            }
        }

        private async Task AddOnConnectedCallbackAsync(Func<Task> handler)
        {
            this.connectedCallbacks.Add(handler);

            if (this.Connected)
            {
                await handler();
            }
        }


        public async Task StartAsync()
        {

            if (this.Started)
            {
                return;
            }

            this.Error = null;
            this.Started = true;
            try
            {
                await OnStatusChanged();
                await this._connection.StartAsync();
                this.Connected = true;

                foreach (var handler in connectedCallbacks)
                {
                    await handler();
                }
            }
            catch (Exception e)
            {
                this.Error = e;
            }

            await OnStatusChanged();
        }

        public async Task StopAsync()
        {
            this.Started = false;
            await OnStatusChanged();
            await this._connection.StopAsync();
        }

        public event Func<Task> StatusChanged;

        protected Task OnStatusChanged()
        {
            var result = StatusChanged?.Invoke();
            return result ?? Task.CompletedTask;
        } 

        public IDisposable On(string eventName, Func<Task> handler)
        {
            return this._connection.On(eventName, handler);
        }

    }
}
